spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -444,6 +444,52 @@ protected void handleMessageInternal(Message<?> message) throws Exception {
 		}
 	}
 
+	protected boolean isExpireGroupsUponCompletion() {
+		return false;
+	}
+
+	private void removeEmptyGroupAfterTimeout(MessageGroup messageGroup, long timeout) {
+		Object groupId = messageGroup.getGroupId();
+		UUID groupUuid = UUIDConverter.getUUID(groupId);
+		ScheduledFuture<?> scheduledFuture = getTaskScheduler()
+				.schedule(() -> {
+					Lock lock = this.lockRegistry.obtain(groupUuid.toString());
+
+					try {
+						lock.lockInterruptibly();
+						try {
+							this.expireGroupScheduledFutures.remove(groupUuid);
+							MessageGroup groupNow = this.messageStore.getMessageGroup(groupUuid);
+							boolean removeGroup = groupNow.getLastModified()
+									<= (System.currentTimeMillis() - this.minimumTimeoutForEmptyGroups);
+							if (removeGroup) {
```

Comment:
Agree about `groupNow.size()` since we can have last modified less than our timeout, but the group has been updated with a new message.

Since the group has been updated it is already that task responsibility to schedule this group for removal on its purpose. Even if we call `scheduledFuture.cancel(true)` it doesn't mean that we exit from this task and a new one can be started already for fresh completed message.

I'll add some comment on the matter here in code, like you did in the `forceComplete()`